txstratum/healthcheck.py

```diff
@@ -0,0 +1,265 @@
+    """This is an interface to be used by other classes implementing health checks for components."""
+
+    @abstractmethod
+    async def get_health_check(self) -> ComponentHealthCheck:
+        """Return the health check status for the component."""
+        raise NotImplementedError()
+
+
+class HealthCheck:
+    """This is the main class that will use the other classes to check the health of the components.
+
+    It will aggregate the responses into a final object to be returned following our standards.
+    """
+
+    def __init__(self, manager: "TxMiningManager", backend: "HathorClient") -> None:
+        """Init the class with the components that will be checked."""
+        self.health_check_components: List[ComponentHealthCheckInterface] = [
+            MiningHealthCheck(manager),
+            FullnodeHealthCheck(backend),
+        ]
+
+    async def get_health_check(self) -> HealthCheckResult:
+        """Return the health check status for the tx-mining-service."""
+        components_health_checks = [
+            await c.get_health_check() for c in self.health_check_components
+        ]
+        components_status = {c.status for c in components_health_checks}
+
+        overall_status = HealthCheckStatus.PASS
+        if HealthCheckStatus.FAIL in components_status:
+            overall_status = HealthCheckStatus.FAIL
+        elif HealthCheckStatus.WARN in components_status:
+            overall_status = HealthCheckStatus.WARN
+
+        return HealthCheckResult(
+            status=overall_status,
+            description="health of tx-mining-service",
+            checks={c.component_name: [c] for c in components_health_checks},
+        )
+
+
+class FullnodeHealthCheck(ComponentHealthCheckInterface):
+    """This class will check the health of the fullnode by sending a request to its /v1a/health."""
+
+    def __init__(self, backend: "HathorClient") -> None:
+        """Init the class with the fullnode backend."""
+        self.backend = backend
+
+    async def get_health_check(self) -> ComponentHealthCheck:
+        """Return the fullnode health check status."""
+        health_check = ComponentHealthCheck(
+            component_name="fullnode",
+            component_type=ComponentType.FULLNODE,
+            # TODO: Ideally we should not use private fields. We'll fix this when fixing line 170
+            component_id=self.backend._base_url,
+            status=HealthCheckStatus.PASS,
+            time=datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ"),
+            output="fullnode is responding correctly",
+        )
+
+        try:
+            # TODO: We need to get the health information from the fullnode, but it's not implemented yet
+            await self.backend.version()
+
+            return health_check
+        except Exception:
+            health_check.status = HealthCheckStatus.FAIL
+            health_check.output = "couldn't connect to fullnode"
+
+            return health_check
```

Comment:
Could we somehow carry the failure information to the health check result? For example, the status code and response content could be helpful in debugging.

Reply:
We will implement a health endpoint for hathor-core as well, which will provide us with detailed information in case the fullnode is unhealthy. This code will be refactor to use this new endpoint when it's ready.

But for now I'll make a change to include the Exception information in the output, which could help in debugging.